This notebook is the scratch paper I used while writing [`evaluate.py`](./evaluate.py). I used this whenever I needed an interactive console to try ideas and view results.

## Load and run `evaluate` module

In [1]:
from importlib import reload # recompile module and re-execute module-level code
import evaluate as ev; reload(ev)

<module 'evaluate' from 'D:\\This PC\\Documents\\Github\\nlp-sandbox\\enron\\evaluate.py'>

In [2]:
%run -i evaluate.py

Importing the model from model.v2.pkl
Loading test data
notes.cleaned.tsv
Featurizing test data
Evaluating

Accuracy: 0.2961992136304063

Confusion matrix:
Layout
[[tn   fp]
 [fn   tp]]

[[ 357 1072]
 [   2   95]]

Classification report:
             precision    recall  f1-score   support

        neg       0.99      0.25      0.40      1429
        pos       0.08      0.98      0.15        97

avg / total       0.94      0.30      0.38      1526


Accuracy (on candidates with 1+ feature): 0.6578073089700996

Confusion matrix:
[[357 204]
 [  2  39]]

Classification report:
             precision    recall  f1-score   support

        neg       0.99      0.64      0.78       561
        pos       0.16      0.95      0.27        41

avg / total       0.94      0.66      0.74       602



**Improvement #1**: accuracy filtering above shows the model should learn to be more uncertain about candidates it has zero features on.
   - This could mean not requiring a VERB in order for a candidate to have features. (Although might be reasonable to scope this problem to tasks of this form first, nail it, then expand the definition)
   - This could mean not including training data with 0 features, with the idea that this will force the model to be more uncertain when it comes across 0-feature candidates in test. But is this a bad bias to introduce?

**Improvement #2**: re-train model using spacy v2

I decided to upgrade spacy in the middle of building this notebook (to use the `displacy` visualizer package later on). Went from `1.9.0` to `2.0.7`. Didn't expect it to affect much until I re-train my model. Wrong.

**Measure**|**v1.9.0**|**v2.0.7 (eval-only)**|**v2.0.7 (re-train)**
-----|-----|-----|-----
Accuracy|0.286|0.312|0.296
Accuracy on 1+ features|0.633|0.698|0.658
Confusion Matrix|[[ 356 1086] [ 4 80]]|[[ 393 1049] [ 1 83]]|[[ 357 1072] [   2   95]]
Confusion Matrix on 1+ features|[[356 216] [ 4 24]]|[[393 181] [ 1 27]]|[[357 204] [  2  39]]

_On the negative side, featurization is slower._

Additional options after upgrade:
- Add [new morphology info](https://spacy.io/api/annotation#pos-tagging) as features (VerbForm, Tense, Aspect)

**TODO**: address label inbalance

## Using: get_sample_predictions

In [3]:
most_incorrect = ev.get_sample_predictions(preds_df, sample_type=ev.SamplePredictions.MOST_INCORRECT, n=20)
most_incorrect

candidate_task  \
1282                                            use the   
529                                               Visit   
1091  end, after court hearings on the issue, the Dr...   
396   ENA Origination Support (Getting info from Tyc...   
1121  late Tuesday accused 29 Enron officers and dir...   
1031                               systematically fail.   
926   wanted only to buy and sell conventional reins...   
167                  Managing Director, Enterprise Risk   
752              Please note your account 00235424-0015   
196                                Published: June 2000   
872   stock options contracts can be hedged with IBM...   
638   Sink out of sight, and the water went over my ...   
923                   and to build balanced portfolios.   
1081       take over the job of administering the plan.   
1305  Certificate Issued: 07/16/2001  96 FERC ? 61,0...   
1176  If you are experiencing problems with today's ...   
1061  Just days before Enron (news/quote) filed for ...   
999   If the facts don't fit the theory, change the ...   
1118  to petition the bankruptcy court to cash in un...   
1058  Enron Paid Out 'Retention' Bonuses Before Bank...   

                      prob_dist label  \
1282  <ProbDist with 2 samples>   neg   
529   <ProbDist with 2 samples>   neg   
1091  <ProbDist with 2 samples>   neg   
396   <ProbDist with 2 samples>   neg   
1121  <ProbDist with 2 samples>   neg   
1031  <ProbDist with 2 samples>   neg   
926   <ProbDist with 2 samples>   neg   
167   <ProbDist with 2 samples>   neg   
752   <ProbDist with 2 samples>   neg   
196   <ProbDist with 2 samples>   neg   
872   <ProbDist with 2 samples>   neg   
638   <ProbDist with 2 samples>   neg   
923   <ProbDist with 2 samples>   neg   
1081  <ProbDist with 2 samples>   neg   
1305  <ProbDist with 2 samples>   neg   
1176  <ProbDist with 2 samples>   neg   
1061  <ProbDist with 2 samples>   neg   
999   <ProbDist with 2 samples>   neg   
1118  <ProbDist with 2 samples>   neg   
1058  <ProbDist with 2 samples>   neg   

                                               features pred  pos_dist  \
1282  {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.963347   
529   {'VERB': 1, 'PARENT_DEP_ROOT': 1, 'PARENT_POS_...  pos  0.960683   
1091  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.955679   
396   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  pos  0.949537   
1121  {'VERB': 1, 'FOLLOWING_POS_NUM': 1, 'FOLLOWING...  pos  0.942597   
1031  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.932591   
926   {'VERB': 3, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...  pos  0.925261   
167   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.902369   
752   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  pos  0.901583   
196   {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.889471   
872   {'VERB': 2, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  pos  0.881920   
638   {'VERB': 2, 'FOLLOWING_POS_ADP': 2, 'FOLLOWING...  pos  0.881467   
923   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  pos  0.870963   
1081  {'VERB': 2, 'FOLLOWING_POS_PART': 1, 'FOLLOWIN...  pos  0.859363   
1305  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.858525   
1176  {'VERB': 4, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  pos  0.857215   
1061  {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.855919   
999   {'VERB': 3, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...  pos  0.854469   
1118  {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.854414   
1058  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.854254   

      neg_dist  max_dist  
1282  0.036653  0.963347  
529   0.039317  0.960683  
1091  0.044321  0.955679  
396   0.050463  0.949537  
1121  0.057403  0.942597  
1031  0.067409  0.932591  
926   0.074739  0.925261  
167   0.097631  0.902369  
752   0.098417  0.901583  
196   0.110529  0.889471  
872   0.118080  0.881920  
638   0.118533  0.881467  
923   0.129037  0.870963  
1081  

In [4]:
ev.print_full_df_column(most_incorrect, ['candidate_task', 'label'], True)

candidate_task	label
use the	neg
Visit	neg
end, after court hearings on the issue, the Drexel estate	neg
ENA Origination Support (Getting info from Tycholiz)--$$XXK	neg
late Tuesday accused 29 Enron officers and directors of	neg
systematically fail.	neg
wanted only to buy and sell conventional reinsurance, Mr. Sweeney	neg
Managing Director, Enterprise Risk	neg
Please note your account 00235424-0015	neg
Published: June 2000	neg
stock options contracts can be hedged with IBM stock, Mr. Lane	neg
Sink out of sight, and the water went over my head.	neg
and to build balanced portfolios.	neg
take over the job of administering the plan.	neg
Certificate Issued: 07/16/2001  96 FERC ? 61,078 (2001)	neg
If you are experiencing problems with today's issue, you may log in to the publisher's site to view the issue here <http://ftenergyusa.com/login_form.asp?FinalURL=/gasdaily/default.asp>. User Name: ENET3318 Password: enron	neg
Just days before Enron (news/quote) filed for bankruptcy	neg
If the fact

### Manual annotations

**candidate\_task**|**label**|**annotation**
:-----:|:-----:|:-----:
ENA Origination Support (Getting info from Tycholiz)--$$XXK|neg|[3(VBG)]
New York Times on the Web, please contact Alyson|neg|[1]
Racer at alyson@nytimes.com or visit our online media|neg|[1]
wanted only to buy and sell conventional reinsurance, Mr. Sweeney|neg|[3(VBD;VB;VB)]
visit http://www.adamsmark.com/resv/rescheck.html.|neg|[1]
freeze: bridge|neg|[1]
use the following link for enrononline: www.enrononline.com|neg|[1]
end, after court hearings on the issue, the Drexel estate|neg|[2]
seek to buy replacement coverage to offset depletions in their|neg|[1][2]
For general information about NYTimes.com, write to|neg|[1][2]
the market observations (as opposed to historical data).|neg|[3(VBN)]
execute the agreement and return it to me via fax no. (713) 646-3490.  I will|neg|[1]
bring these issues to the table, in a meeting involving all the concerned parties|neg|[1]
late Tuesday accused 29 Enron officers and directors of|neg|[2][3(VBD)]
John Elder, creator of a grass-roots investment program to develop|neg|[2][3(VB)]
Get FREE shipping on orders of $75 or more at Starbucks.com|neg|[1]
Managing Director, Enterprise Risk|neg|[3(VBG)]
use the|neg|[1][2]
2. Decision support system: a model recommending transactions to a trader without|neg|[3(VBG)]
convince the external parties (stock analysts, creditors,  credit rating agencies) about|neg|[1]

**Annotation Key**:
- [1] = Reasonable task
- [2] = Sentence fragment
- [3(\*)] = Overweighted VERB (VB\* forms)


([Markdown Table Generator](https://jakebathman.github.io/Markdown-Table-Generator/))

#### Improvements
- **#3** Add feature around the specific kind of VERB (present vs past, etc)
- **#4** Might be useful to classify the greater body of text (e.g., article, paragraph, ?) as something worth further parsing by line - to help avoid sentence fragmentation or pulling "reasonable tasks" out of irrelevant text.

In [5]:
ev.render_displacy(most_incorrect, ev.DisplacyStyles.dep)

['VB', 'DT']
-----



['VB']
-----



['VB', ',', 'IN', 'NN', 'NNS', 'IN', 'DT', 'NN', ',', 'DT', 'NNP', 'NN']
-----



['NNP', 'NNP', 'NNP', '-LRB-', 'VBG', 'NN', 'IN', 'NN']
-----



['JJ', 'NNP', 'VBD', 'CD', 'NNP', 'NNS', 'CC', 'NNS', 'IN']
-----



['RB', 'VB', '.']
-----



['VBD', 'RB', 'TO', 'VB', 'CC', 'VB', 'JJ', 'NN', ',', 'NNP', 'NNP']
-----



['VBG', 'NNP', ',', 'NN', 'NN']
-----



['UH', 'VB', 'PRP$', 'NN', 'CD', 'HYPH', 'CD']
-----



['VBN', ':', 'NNP', 'CD']
-----



['NN', 'NNS', 'NNS', 'MD', 'VB', 'VBN', 'IN', 'NNP', 'NN', ',', 'NNP', 'NNP']
-----



['VB', 'IN', 'IN', 'NN', ',', 'CC', 'DT', 'NN', 'VBD', 'IN', 'PRP$', 'NN', '.']
-----



['CC', 'TO', 'VB', 'JJ', 'NNS', '.']
-----



['VB', 'RP', 'DT', 'NN', 'IN', 'VBG', 'DT', 'NN', '.']
-----



['NNP', 'VBD', ':', 'CD', '', 'CD', 'NN', '.', 'CD', '-LRB-', 'CD', '-RRB-']
-----



['IN', 'PRP', 'VBP', 'VBG', 'NNS', 'IN', 'NN', 'POS', 'NN', ',', 'PRP', 'MD', 'VB', 'RP', 'IN', 'DT', 'NN', 'POS', 'NN', 'TO', 'VB', 'DT', 'NN', 'RB', 'XX', 'ADD', 'SYM', 'RB', 'SYM', 'NN', 'NNP', 'NNP', ':', 'NNP', 'NNP', ':', 'NN']
-----



['RB', 'NNS', 'IN', 'NNP', '-LRB-', 'NN', 'SYM', 'NN', '-RRB-', 'VBD', 'IN', 'NN']
-----



['IN', 'DT', 'NNS', 'VBP', 'RB', 'VB', 'DT', 'NN', ',', 'VB', 'DT', 'NNS', '.']
-----



['TO', 'VB', 'DT', 'NN', 'NN', 'TO', 'VB', 'IN', 'JJ', 'NN']
-----



['NNP', 'VBD', 'NNP', '``', 'NNP', "''", 'NNS', 'IN', 'NNP', 'NNP']
-----



In [6]:
most_incorrect_pos = ev.get_sample_predictions(preds_df, sample_type=ev.SamplePredictions.MOST_INCORRECT_POS, n=20)
most_incorrect_pos

candidate_task                  prob_dist  \
380  maine impossible to get to...next option?  <ProbDist with 2 samples>   
191         Start project  with Elene for Mark  <ProbDist with 2 samples>   

    label                                           features pred  pos_dist  \
380   pos  {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  neg  0.330498   
191   pos  {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  0.480333   

     neg_dist  max_dist  
380  0.669502  0.669502  
191  0.519667  0.519667

In [7]:
most_correct = ev.get_sample_predictions(preds_df, sample_type=ev.SamplePredictions.MOST_CORRECT, n=20)
most_correct

candidate_task  \
1215                              Call Grant about Tony   
762                                           Bring AER   
33    visit http://www.adamsmark.com/resv/rescheck.h...   
1255                     Call Ambrosia ? Monday morning   
1045  Another aspect of this problem is prioritizati...   
121   35. "Stewardesses" is the longest word that is...   
83    5. A shark is the only fish that can blink wit...   
88                        building is an American flag.   
990   Common sense is the collection of prejudices a...   
163                                        Check for ML   
620   Like an eye between two white lids that will n...   
109   cop and Ernie the taxi driver in Frank Capra's...   
696                                  Send check to Lisa   
459   800   357  4410  ext  6422  Amy regardinding T...   
612    The tulips are too excitable, it is winter here.   
996   The release of atomic energy has not created a...   
1294  There were reports of further attacks overnigh...   
1267  The old man steps up to the tee and hits the b...   
411                                   Get Kerri a gift.   
1328                                     freeze: bridge   

                      prob_dist label  \
1215  <ProbDist with 2 samples>   pos   
762   <ProbDist with 2 samples>   pos   
33    <ProbDist with 2 samples>   pos   
1255  <ProbDist with 2 samples>   pos   
1045  <ProbDist with 2 samples>   neg   
121   <ProbDist with 2 samples>   neg   
83    <ProbDist with 2 samples>   neg   
88    <ProbDist with 2 samples>   neg   
990   <ProbDist with 2 samples>   neg   
163   <ProbDist with 2 samples>   pos   
620   <ProbDist with 2 samples>   neg   
109   <ProbDist with 2 samples>   neg   
696   <ProbDist with 2 samples>   pos   
459   <ProbDist with 2 samples>   pos   
612   <ProbDist with 2 samples>   neg   
996   <ProbDist with 2 samples>   neg   
1294  <ProbDist with 2 samples>   neg   
1267  <ProbDist with 2 samples>   neg   
411   <ProbDist with 2 samples>   pos   
1328  <ProbDist with 2 samples>   pos   

                                               features pred  pos_dist  \
1215  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.996769   
762   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.995642   
33    {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
1255  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
1045  {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  0.013304   
121   {'VERB': 3, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.015950   
83    {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.016720   
88    {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.019714   
990   {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.022955   
163   {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.973474   
620   {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  neg  0.027507   
109   {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.028090   
696   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  pos  0.971138   
459   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.969010   
612   {'VERB': 2, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...  neg  0.031746   
996   {'VERB': 6, 'FOLLOWING_POS_ADV': 2, 'FOLLOWING...  neg  0.033124   
1294  {'VERB': 4, 'FOLLOWING_POS_NOUN': 2, 'FOLLOWIN...  neg  0.033351   
1267  {'VERB': 3, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  neg  0.033445   
411   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.964902   
1328  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.964522   

      neg_dist  max_dist  
1215  0.003231  0.996769  
762   0.004358  0.995642  
33    0.008315  0.991685  
1255  0.008315  0.991685  
1045  0.986696  0.986696  
121   0.984050  0.984050  
83    0.983280  0.983280  
88    0.980286  0.980286  
990   0.977045  0.977045  
163   0.026526  0.973474  
620   0.972493  0.972493  
109   0.971910  0.971910  
696   0.028862  0.971138  
459   

In [8]:
ev.print_full_df_column(most_correct, ['candidate_task', 'label'], True)

candidate_task	label
Call Grant about Tony	pos
Bring AER	pos
visit http://www.adamsmark.com/resv/rescheck.html.	pos
Call Ambrosia ? Monday morning	pos
Another aspect of this problem is prioritization of different projects. In my view	neg
"35. ""Stewardesses"" is the longest word that is typed with only the left"	neg
5. A shark is the only fish that can blink with both eyes.	neg
building is an American flag.	neg
Common sense is the collection of prejudices acquired by age eighteen.	neg
Check for ML	pos
Like an eye between two white lids that will not shut.	neg
"cop and Ernie the taxi driver in Frank Capra's ""It's a Wonderful Life."""	neg
Send check to Lisa	pos
800   357  4410  ext  6422  Amy regardinding Telefonica	pos
The tulips are too excitable, it is winter here.	neg
"The release of atomic energy has not created a new problem. It has	neg
merely made more urgent the necessity of solving an existing one."	neg
There were reports of further attacks overnight on the Kajaki Dam area, whe

In [9]:
most_correct_pos = ev.get_sample_predictions(preds_df, sample_type=ev.SamplePredictions.MOST_CORRECT_POS, n=20)
most_correct_pos

candidate_task  \
1215                              Call Grant about Tony   
762                                           Bring AER   
33    visit http://www.adamsmark.com/resv/rescheck.h...   
1255                     Call Ambrosia ? Monday morning   
163                                        Check for ML   
696                                  Send check to Lisa   
459   800   357  4410  ext  6422  Amy regardinding T...   
411                                   Get Kerri a gift.   
1328                                     freeze: bridge   
1016  convince the external parties (stock analysts,...   
1137  For general information about NYTimes.com, wri...   
1055  Get FREE shipping on orders of $75 or more at ...   
605                                 Send ED-mail to Ed.   
1281           use the following http:\\enrononline.com   
273   Check last years tax return for RRSP contribut...   
1280  use the following link for enrononline: www.en...   
1134   New York Times on the Web, please contact Alyson   
75                      resend pheasant email to all.df   
1516            please download and make a backup copy.   
1517  install v4.0 build 23, follow old instructions...   

                      prob_dist label  \
1215  <ProbDist with 2 samples>   pos   
762   <ProbDist with 2 samples>   pos   
33    <ProbDist with 2 samples>   pos   
1255  <ProbDist with 2 samples>   pos   
163   <ProbDist with 2 samples>   pos   
696   <ProbDist with 2 samples>   pos   
459   <ProbDist with 2 samples>   pos   
411   <ProbDist with 2 samples>   pos   
1328  <ProbDist with 2 samples>   pos   
1016  <ProbDist with 2 samples>   pos   
1137  <ProbDist with 2 samples>   pos   
1055  <ProbDist with 2 samples>   pos   
605   <ProbDist with 2 samples>   pos   
1281  <ProbDist with 2 samples>   pos   
273   <ProbDist with 2 samples>   pos   
1280  <ProbDist with 2 samples>   pos   
1134  <ProbDist with 2 samples>   pos   
75    <ProbDist with 2 samples>   pos   
1516  <ProbDist with 2 samples>   pos   
1517  <ProbDist with 2 samples>   pos   

                                               features pred  pos_dist  \
1215  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.996769   
762   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.995642   
33    {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
1255  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
163   {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.973474   
696   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  pos  0.971138   
459   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.969010   
411   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.964902   
1328  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.964522   
1016  {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.961552   
1137  {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.961096   
1055  {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  pos  0.957283   
605   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  pos  0.954658   
1281  {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.947790   
273   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  pos  0.941247   
1280  {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.941086   
1134  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.928820   
75    {'VERB': 2, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  pos  0.926999   
1516  {'VERB': 2, 'FOLLOWING_POS_CCONJ': 1, 'FOLLOWI...  pos  0.922628   
1517  {'VERB': 4, 'FOLLOWING_POS_NOUN': 3, 'FOLLOWIN...  pos  0.920888   

      neg_dist  max_dist  
1215  0.003231  0.996769  
762   0.004358  0.995642  
33    0.008315  0.991685  
1255  0.008315  0.991685  
163   0.026526  0.973474  
696   0.028862  0.971138  
459   0.030990  0.969010  
411   0.035098  0.964902  
1328  0.035478  0.964522  
1016  0.038448  0.961552  
1137  0.038904  0.961096  
1055  0.042717  0.957283  
605   0.045342  0.954658  
1281  

In [10]:
most_uncertain = ev.get_sample_predictions(preds_df, sample_type=ev.SamplePredictions.MOST_UNCERTAIN, n=20)
most_uncertain

candidate_task  \
965   Currently, most of the trades are between secu...   
573   Davis has taken heat for locking up too much e...   
614   I am learning peacefulness, lying by myself qu...   
721   about Australian markets and was greatly impre...   
1411  Incentive fee (in main contract) defined to on...   
1317  trader I see an opportunity to sell soybeans a...   
901   The oldest of the insurance-related, exchange-...   
116   31. The microwave was invented after a researc...   
904   Initially, there was substantial interest in t...   
909   so far this year, he said. Last year, there wa...   
991   I know not with what weapons World War III wil...   
1100  If creditors can convince the court that the e...   
853   natural catastrophes, the allure of the exchan...   
812   fees charged per transaction. My comment: in c...   
652   Upsetting me with their sudden tongues and the...   
1304                                  Filed: 03/29/2001   
710                                    10800 Gosling RD   
191                  Start project  with Elene for Mark   
441               Rafael  Pina @ shell is Lia's Contact   
1041  for the model calibration. Right now this resp...   

                      prob_dist label  \
965   <ProbDist with 2 samples>   neg   
573   <ProbDist with 2 samples>   neg   
614   <ProbDist with 2 samples>   neg   
721   <ProbDist with 2 samples>   neg   
1411  <ProbDist with 2 samples>   neg   
1317  <ProbDist with 2 samples>   neg   
901   <ProbDist with 2 samples>   neg   
116   <ProbDist with 2 samples>   neg   
904   <ProbDist with 2 samples>   neg   
909   <ProbDist with 2 samples>   neg   
991   <ProbDist with 2 samples>   neg   
1100  <ProbDist with 2 samples>   neg   
853   <ProbDist with 2 samples>   neg   
812   <ProbDist with 2 samples>   neg   
652   <ProbDist with 2 samples>   neg   
1304  <ProbDist with 2 samples>   neg   
710   <ProbDist with 2 samples>   neg   
191   <ProbDist with 2 samples>   pos   
441   <ProbDist with 2 samples>   neg   
1041  <ProbDist with 2 samples>   neg   

                                               features pred  pos_dist  \
965   {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.500492   
573   {'VERB': 6, 'FOLLOWING_POS_VERB': 3, 'FOLLOWIN...  neg  0.498548   
614   {'VERB': 3, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  pos  0.501745   
721   {'VERB': 1, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...  neg  0.496814   
1411  {'VERB': 2, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.504848   
1317  {'VERB': 5, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.505635   
901   {'VERB': 2, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  neg  0.492510   
116   {'VERB': 3, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  neg  0.491751   
904   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  neg  0.491579   
909   {'VERB': 3, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.512387   
991   {'VERB': 5, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...  pos  0.513986   
1100  {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.485806   
853   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  pos  0.514800   
812   {'VERB': 2, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.516990   
652   {'VERB': 1, 'FOLLOWING_POS_PRON': 1, 'FOLLOWIN...  pos  0.519250   
1304  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  neg  0.480433   
710   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  0.480333   
191   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  0.480333   
441   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  neg  0.479154   
1041  {'VERB': 2, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  neg  0.478514   

      neg_dist  max_dist  
965   0.499508  0.500492  
573   0.501452  0.501452  
614   0.498255  0.501745  
721   0.503186  0.503186  
1411  0.495152  0.504848  
1317  0.494365  0.505635  
901   0.507490  0.507490  
116   0.508249  0.508249  
904   0.508421  0.508421  
909   0.487613  0.512387  
991   0.486014  0.513986  
1100  0.514194  0.514194  
853   0.485200  0.514800  
812   

In [11]:
ev.print_full_df_column(most_uncertain, ['candidate_task', 'pred', 'label'], True)

candidate_task	pred	label
Currently, most of the trades are between securities dealers	pos	neg
Davis has taken heat for locking up too much electricity at high prices. Similarly, the Department of Water Resources has been criticized for amateurish purchasing practices.	neg	neg
I am learning peacefulness, lying by myself quietly	pos	neg
about Australian markets and was greatly impressed with the	neg	neg
Incentive fee (in main contract) defined to only include the sale of gas.	pos	neg
trader I see an opportunity to sell soybeans and buy corn. I am thinking	pos	neg
The oldest of the insurance-related, exchange-based derivative	neg	neg
31. The microwave was invented after a researcher walked by radar tube and a	neg	neg
Initially, there was substantial interest in the options, but the soft	neg	neg
so far this year, he said. Last year, there was increased interest in	pos	neg
"I know not with what weapons World War III will be fought, but World	neg
War IV will be fought with sticks and stones

In [12]:
ev.confusion_matrix(most_uncertain[['candidate_task', 'prob_dist', 'label', 'features']].itertuples(index=False))

array([[10,  9],
       [ 1,  0]])

#### Q: What TP candidates (actual tasks) didn't have features?

In [13]:
all_correct_pos = ev.get_sample_predictions(preds_df, sample_type=ev.SamplePredictions.MOST_CORRECT_POS)

In [14]:
len(all_correct_pos)

95

In [15]:
no_features_mask = all_correct_pos['features'] == {}
tp_no_features = all_correct_pos.iloc[np.where(no_features_mask)]
tp_no_features

candidate_task                  prob_dist label  \
984                               Headcount  <ProbDist with 2 samples>   pos   
1253                                     RT  <ProbDist with 2 samples>   pos   
764                              10:00  Sat  <ProbDist with 2 samples>   pos   
1485                   Monday at 11:30 Tara  <ProbDist with 2 samples>   pos   
1252                                Thanks,  <ProbDist with 2 samples>   pos   
755                      3. Barbara Weather  <ProbDist with 2 samples>   pos   
754                                2. Accom  <ProbDist with 2 samples>   pos   
753                              1, Li Xiao  <ProbDist with 2 samples>   pos   
747                 Article about the group  <ProbDist with 2 samples>   pos   
746                               McCormack  <ProbDist with 2 samples>   pos   
1234                            exact time.  <ProbDist with 2 samples>   pos   
1171              EES presentations to Rhys  <ProbDist with 2 samples>   pos   
604               Dentist on  Jan 25, 7:45.  <ProbDist with 2 samples>   pos   
602               Dentist on  Jan 25, 7:45.  <ProbDist with 2 samples>   pos   
1172                      Krishna's letters  <ProbDist with 2 samples>   pos   
8                              speed on eol  <ProbDist with 2 samples>   pos   
597                                Articles  <ProbDist with 2 samples>   pos   
596                                   Vince  <ProbDist with 2 samples>   pos   
286                    $933  to Dean Witter  <ProbDist with 2 samples>   pos   
276                 - MYERS, JENNIFER M.doc  <ProbDist with 2 samples>   pos   
275                  - HUSSER, SHANNA L.doc  <ProbDist with 2 samples>   pos   
274                     - FLINN, DARA M.doc  <ProbDist with 2 samples>   pos   
184                                $4676.18  <ProbDist with 2 samples>   pos   
183                                To Keith  <ProbDist with 2 samples>   pos   
182                                 $20,709  <ProbDist with 2 samples>   pos   
181                                To Keith  <ProbDist with 2 samples>   pos   
180                                 3.Steve  <ProbDist with 2 samples>   pos   
179                          2. John Martin  <ProbDist with 2 samples>   pos   
178                         1. Presentation  <ProbDist with 2 samples>   pos   
177          Call John Martin Sat 3:00 p.m.  <ProbDist with 2 samples>   pos   
162                             Phone cards  <ProbDist with 2 samples>   pos   
128                 files into that folder.  <ProbDist with 2 samples>   pos   
126                                  Danny,  <ProbDist with 2 samples>   pos   
76                                     memo  <ProbDist with 2 samples>   pos   
11                        dave or tom moran  <ProbDist with 2 samples>   pos   
341                    Outline for Stinson.  <ProbDist with 2 samples>   pos   
378   Power 2000 Operational Risk for Kevin  <ProbDist with 2 samples>   pos   
379                                rent tix  <ProbDist with 2 samples>   pos   
485                   1. Paper fro Kevin K.  <ProbDist with 2 samples>   pos   
527                              1. M. Soto  <ProbDist with 2 samples>   pos   
499                            1. Ecopetrol  <ProbDist with 2 samples>   pos   
490                                  Piazze  <ProbDist with 2 samples>   pos   
489                                Johansen  <ProbDist with 2 samples>   pos   
9                        message for monday  <ProbDist with 2 samples>   pos   
487                            Andy kalotay  <ProbDist with 2 samples>   pos   
486                    2. Rosemary  at ICBI  <ProbDist with 2 samples>   pos   
483                                  Lacima  <ProbDist with 2 samples>   pos   
409                           Jean Cummings  <ProbDist with 2 samples>   pos   
482                             Nelson Neal  <ProbDist with 2 samples>   pos   
481                          

**Improvement #5**: Manual re-labeling due to my "split by new line" decision is necessary for more accurate results. All sample datasets have instances of mis-labeled data.

In [16]:
ev.render_displacy(tp_no_features, [ev.DisplacyStyles.dep, ev.DisplacyStyles.ent])

['NNP']
-----



['NNP']
-----



['CD', '', 'NNP']
-----



['NNP', 'IN', 'CD', 'NN']
-----



['NNS', ',']
-----



['LS', '.', 'NNP', 'NNP']
-----



['LS', '.', 'ADD']
-----



['LS', ',', 'NNP', 'NNP']
-----



['NN', 'IN', 'DT', 'NN']
-----



['NN']
-----



['JJ', 'NN', '.']
-----



['NNP', 'NNS', 'IN', 'NNP']
-----



['NN', 'IN', '', 'NNP', 'CD', ',', 'CD', '.']
-----



['NN', 'IN', '', 'NNP', 'CD', ',', 'CD', '.']
-----



['NNP', 'POS', 'NNS']
-----



['NN', 'IN', 'NN']
-----



['NNS']
-----



['NN']
-----



['$', 'CD', '', 'IN', 'NNP', 'NNP']
-----



[':', 'NNP', ',', 'NNP', 'NNS']
-----



[':', 'NNP', ',', 'NNP', 'NNP']
-----



['HYPH', 'NNP', ',', 'NNP', 'NNP']
-----



['$', 'CD']
-----



['IN', 'NNP']
-----



['$', 'CD']
-----



['IN', 'NNP']
-----



['CD']
-----



['LS', '.', 'NNP', 'NNP']
-----



['LS', '.', 'NN']
-----



['NNP', 'NNP', 'NNP', 'NNP', 'CD', 'NN']
-----



['NN', 'NNS']
-----



['NNS', 'IN', 'DT', 'NN', '.']
-----



['NNP', ',']
-----



['NN']
-----



['NN', 'CC', 'NN', 'NN']
-----



['NNP', 'IN', 'NNP', '.']
-----



['NN', 'CD', 'NNP', 'NN', 'IN', 'NNP']
-----



['NN', 'NN']
-----



['LS', '.', 'NN', 'NN', 'NNP', 'NNP']
-----



['LS', '.', 'NNP', 'NNP']
-----



['LS', '.', 'NN']
-----



['NNP']
-----



['NNP']
-----



['NN', 'IN', 'NN']
-----



['NNP', 'NN']
-----



['LS', '.', 'NNP', '', 'IN', 'RB']
-----



['NNP']
-----



['NNP', 'NNP']
-----



['NNP', 'NNP']
-----



['NNP']
-----



['JJ', 'HYPH', 'NN']
-----



['NN']
-----



['NNP', '', 'CD', 'NN', '_SP', 'PRP', 'MD']
-----



['NN', 'IN', 'NNP']
-----



['CD', 'SYM', 'CD']
-----



['NNP', 'NNP']
-----



### Observations

- Model is bad at:
    - Incomplete sentences/fragments
    - Tasks without verbs (noun phrases)
    - Taking verb tense into account
- Model is good at:
    - Finding tasks in weird contexts
- Model is most uncertain about: 

## Writing: get_sample_predictions

Understanding code from [fast.ai 2018: lesson 1](https://github.com/fastai/fastai/blob/master/courses/dl1/lesson1.ipynb) for getting various types of sample predictions.

`def rand_by_mask(mask, n): return np.random.choice(np.where(mask)[0], n, replace=False)`

    np.random.choice(a, size=None, replace=True, p=None)

    np.where(condition[, x, y])

`def rand_by_correct(preds, is_correct): return rand_by_mask((preds.preds == preds.labels)==is_correct)`
    
    np.where((preds.preds == preds.labels)==is_correct))

Goals:
1. Get `preds` into a form so that `rand_by_mask` works on it
2. Get `preds` into a form so that we can find what is correct vs incorrect

### What is `preds`?

In [17]:
type(preds)

list

In [18]:
type(preds[1]), preds[1]

(tuple,
 ('http://gasfundy.corp.enron.com/gas/framework/default.asp',
  <ProbDist with 2 samples>,
  'neg',
  {}))

where `tuple == (candidate_task, prediction_dist, label, features)`

In [19]:
[(p[0], p[1].max(), p[2], p[3]) for p in preds][1]

('http://gasfundy.corp.enron.com/gas/framework/default.asp', 'pos', 'neg', {})

In [20]:
type(preds[1][1]), preds[1][1]

(nltk.probability.DictionaryProbDist, <ProbDist with 2 samples>)

In [21]:
import nltk
??nltk.probability.DictionaryProbDist

In [22]:
preds[1][1].samples()

dict_keys(['neg', 'pos'])

In [23]:
preds[1][1].max()

'pos'

In [24]:
preds[1][1].prob('pos'), preds[1][1].prob('neg')

(0.52202910464797891, 0.47797089535202103)

### Form of `preds` that works with `rand_by_mask(mask, n)`?

For `np.random.choice`, first arg must be 1-dimensional

In [25]:
import numpy as np
np_preds = np.array(preds)
np_preds.shape

(1526, 4)

`preds` as it stands will not work. Instead, we can feed in the length of `preds` and let `np.random.choice` select random indices of `preds`.

In [26]:
n = 5 # number of results
mask = len(np_preds)

In [27]:
np_random = np_preds[np.random.choice(mask, n)]

In [28]:
np_random

array([['http://www.science-finance.fr/publications.html',
        <ProbDist with 2 samples>, 'neg', {}],
       ['They pass the way gulls pass inland in their white caps,',
        <ProbDist with 2 samples>, 'neg',
        {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING_POSTAG_DT': 1, 'PARENT_DEP_ROOT': 1, 'PARENT_POS_VERB': 1, 'PARENT_POSTAG_VBP': 1, 'CHILD_DEP_NSUBJ': 2, 'CHILD_POS_PRON': 1, 'CHILD_POSTAG_PRP': 1, 'CHILD_DEP_DOBJ': 1, 'CHILD_POS_NOUN': 2, 'CHILD_POSTAG_NN': 1, 'CHILD_DEP_PUNCT': 1, 'CHILD_POS_PUNCT': 1, 'CHILD_POSTAG_,': 1, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING_POSTAG_RB': 1, 'PARENT_DEP_DOBJ': 1, 'PARENT_POS_NOUN': 1, 'PARENT_POSTAG_NN': 1, 'CHILD_POSTAG_NNS': 1, 'CHILD_DEP_ADVMOD': 1, 'CHILD_POS_ADV': 1, 'CHILD_POSTAG_RB': 1, 'CHILD_DEP_PREP': 1, 'CHILD_POS_ADP': 1, 'CHILD_POSTAG_IN': 1}],
       ['Houston Chronicle Online', <ProbDist with 2 samples>, 'neg', {}],
       ['ID:         mcconnellmarkr', <ProbDist with 2 samples>, 'neg', {}],
       ['https://www.admissio

In [29]:
len(np_random)

5

### Form of `preds` that works with `rand_by_correct(preds, is_correct)`?

`preds` must be broken down into predictions and labels. So first, let's get the predictions.

Remember that each `tuple` in `preds` == `(candidate_task, prediction_dist, label, features)`

In [30]:
[p[1].max() for p in preds][:n]

['pos', 'pos', 'pos', 'pos', 'neg']

Labels can be retrieved similarly

In [31]:
[p[2] for p in preds][:n]

['neg', 'neg', 'neg', 'neg', 'neg']

Noticing that `preds` might be more readable as a pandas dataframe. So let's convert it.

In [32]:
import pandas as pd

preds_df = pd.DataFrame.from_records(preds, columns=['candidate_task', 'prob_dist', 'label', 'features'])
preds_df['pred'] = [p.max() for p in preds_df['prob_dist']]

In [33]:
preds_df[:n]

candidate_task  \
0                      https://www4.rsweb.com/61045/   
1  http://gasfundy.corp.enron.com/gas/framework/d...   
2                                       212 836 5030   
3        http://fundamentals.corp.enron.com/main.asp   
4                                      adams 30  631   

                   prob_dist label  \
0  <ProbDist with 2 samples>   neg   
1  <ProbDist with 2 samples>   neg   
2  <ProbDist with 2 samples>   neg   
3  <ProbDist with 2 samples>   neg   
4  <ProbDist with 2 samples>   neg   

                                            features pred  
0                                                 {}  pos  
1                                                 {}  pos  
2                                                 {}  pos  
3                                                 {}  pos  
4  {'VERB': 1, 'FOLLOWING_POS_NUM': 1, 'FOLLOWING...  neg

In [34]:
[p.max() for p in preds_df['prob_dist']][:n]

['pos', 'pos', 'pos', 'pos', 'neg']

In [35]:
preds_df['label'][:n].values

array(['neg', 'neg', 'neg', 'neg', 'neg'], dtype=object)

Put this all together to write our own 

#### `rand_by_correct(preds, n=5, is_correct=True)`

In [36]:
def rand_by_mask(mask, n):
    return np.random.choice(np.where(mask)[0], n, replace=False)

def rand_by_correct(preds_df, n = 5, is_correct = True):
    return preds_df.iloc[rand_by_mask(([p.max() for p in preds_df['prob_dist']] == preds_df['label'].values)==is_correct, n)]

In [37]:
# correct
rand_by_correct(preds_df)

candidate_task  \
15    4) What happens to the Restricted Stock grant ...   
787   and locations in the Houston area.  Choose the...   
1064                 who had dealings with the company.   
1152  our website,I would like you to chair one of t...   
966   themselves, but, eventually, the contracts wil...   

                      prob_dist label  \
15    <ProbDist with 2 samples>   neg   
787   <ProbDist with 2 samples>   neg   
1064  <ProbDist with 2 samples>   neg   
1152  <ProbDist with 2 samples>   neg   
966   <ProbDist with 2 samples>   neg   

                                               features pred  
15    {'VERB': 9, 'FOLLOWING_POS_ADP': 2, 'FOLLOWING...  neg  
787   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  neg  
1064  {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  
1152  {'VERB': 1, 'FOLLOWING_POS_PRON': 1, 'FOLLOWIN...  neg  
966   {'VERB': 2, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  neg

In [38]:
# incorrect
rand_by_correct(preds_df, is_correct=False)

candidate_task  \
1111  payments as necessary "in order to protect and...   
470   PaineWebber (713) 654 0371 Rocky Emery        ...   
606               http://www.evite.com/reminders?li=reg   
798                                               Greg,   
191                  Start project  with Elene for Mark   

                      prob_dist label  \
1111  <ProbDist with 2 samples>   neg   
470   <ProbDist with 2 samples>   neg   
606   <ProbDist with 2 samples>   neg   
798   <ProbDist with 2 samples>   neg   
191   <ProbDist with 2 samples>   pos   

                                               features pred  
1111  {'VERB': 2, 'FOLLOWING_POS_CCONJ': 1, 'FOLLOWI...  pos  
470                                                  {}  pos  
606                                                  {}  pos  
798                                                  {}  pos  
191   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg

Cool! Adding more goals... this time around getting the 'most correct/incorrect'

`def most_by_mask(mask, mult):
    idxs = np.where(mask)[0]
    return idxs[np.argsort(mult * probs[idxs])[:4]]`
    
    numpy.argsort(a, axis=-1, kind='quicksort', order=None)

`def most_by_correct(y, is_correct): 
    mult = -1 if (y==1)==is_correct else 1
    return most_by_mask(((preds == data.val_y)==is_correct) & (data.val_y == y), mult)`
    
    mult: makes it so that if 'correct' == 'lowest probabilities', the argsort sees lower == more correct. else, higher == more correct

Goals:
1. Get `preds` into a form so that `most_by_mask` works on it
2. Get `preds` into a form so that we can find what is _most_ correct vs _most_ incorrect

Unlike the `Cats vs Dogs` predictions, NLTK provides separate probability distributions for each possible label.

Sorting is easy if we break up the values within the `prob_dist` column

In [39]:
[p.prob('pos') for p in preds_df['prob_dist']][:n]

[0.52202910464797891,
 0.52202910464797891,
 0.52202910464797891,
 0.52202910464797891,
 0.33223657166405279]

In [40]:
[p.prob('neg') for p in preds_df['prob_dist']][:n]

[0.47797089535202103,
 0.47797089535202103,
 0.47797089535202103,
 0.47797089535202103,
 0.66776342833594737]

In [41]:
preds_df['pos_dist'] = [p.prob('pos') for p in preds_df['prob_dist']]
preds_df['neg_dist'] = [p.prob('neg') for p in preds_df['prob_dist']]

In [42]:
preds_df[:n]

candidate_task  \
0                      https://www4.rsweb.com/61045/   
1  http://gasfundy.corp.enron.com/gas/framework/d...   
2                                       212 836 5030   
3        http://fundamentals.corp.enron.com/main.asp   
4                                      adams 30  631   

                   prob_dist label  \
0  <ProbDist with 2 samples>   neg   
1  <ProbDist with 2 samples>   neg   
2  <ProbDist with 2 samples>   neg   
3  <ProbDist with 2 samples>   neg   
4  <ProbDist with 2 samples>   neg   

                                            features pred  pos_dist  neg_dist  
0                                                 {}  pos  0.522029  0.477971  
1                                                 {}  pos  0.522029  0.477971  
2                                                 {}  pos  0.522029  0.477971  
3                                                 {}  pos  0.522029  0.477971  
4  {'VERB': 1, 'FOLLOWING_POS_NUM': 1, 'FOLLOWING...  neg  0.332237  0.667763

In [43]:
preds_df.sort_values('pos_dist', ascending=False)[:5]

candidate_task  \
1215                              Call Grant about Tony   
762                                           Bring AER   
1255                     Call Ambrosia ? Monday morning   
33    visit http://www.adamsmark.com/resv/rescheck.h...   
163                                        Check for ML   

                      prob_dist label  \
1215  <ProbDist with 2 samples>   pos   
762   <ProbDist with 2 samples>   pos   
1255  <ProbDist with 2 samples>   pos   
33    <ProbDist with 2 samples>   pos   
163   <ProbDist with 2 samples>   pos   

                                               features pred  pos_dist  \
1215  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.996769   
762   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.995642   
1255  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
33    {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
163   {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.973474   

      neg_dist  
1215  0.003231  
762   0.004358  
1255  0.008315  
33    0.008315  
163   0.026526

In [44]:
preds_df.sort_values('neg_dist', ascending=False)[:5]

candidate_task  \
1045  Another aspect of this problem is prioritizati...   
121   35. "Stewardesses" is the longest word that is...   
83    5. A shark is the only fish that can blink wit...   
88                        building is an American flag.   
990   Common sense is the collection of prejudices a...   

                      prob_dist label  \
1045  <ProbDist with 2 samples>   neg   
121   <ProbDist with 2 samples>   neg   
83    <ProbDist with 2 samples>   neg   
88    <ProbDist with 2 samples>   neg   
990   <ProbDist with 2 samples>   neg   

                                               features pred  pos_dist  \
1045  {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  0.013304   
121   {'VERB': 3, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.015950   
83    {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.016720   
88    {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.019714   
990   {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.022955   

      neg_dist  
1045  0.986696  
121   0.984050  
83    0.983280  
88    0.980286  
990   0.977045

In [45]:
preds[1][1].samples()

dict_keys(['neg', 'pos'])

Drumroll...

#### most_by_correct(preds_df, label, n=5, is_correct=True)

In [46]:
def most_by_correct(preds_df, label, n=5, is_correct=True):
    label_options = list(preds_df['prob_dist'][0].samples())
    # note the 2-class assumption
    other_label = label_options[abs(label_options.index(label)-1)]
    # if is_correct, label dist is what matters; if incorrect, look at other_label dist
    dist_column = label + '_dist' if is_correct == True else other_label + '_dist'
    
    correct_mask = ([p.max() for p in preds_df['prob_dist']] == preds_df['label'].values)==is_correct
    target_label_mask = preds_df['label'].values == label
    
    return preds_df.iloc[np.where(correct_mask & target_label_mask)].sort_values(dist_column, ascending=False)[:n]

In [47]:
most_by_correct(preds_df, 'pos')

candidate_task  \
1215                              Call Grant about Tony   
762                                           Bring AER   
33    visit http://www.adamsmark.com/resv/rescheck.h...   
1255                     Call Ambrosia ? Monday morning   
163                                        Check for ML   

                      prob_dist label  \
1215  <ProbDist with 2 samples>   pos   
762   <ProbDist with 2 samples>   pos   
33    <ProbDist with 2 samples>   pos   
1255  <ProbDist with 2 samples>   pos   
163   <ProbDist with 2 samples>   pos   

                                               features pred  pos_dist  \
1215  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.996769   
762   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.995642   
33    {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
1255  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...  pos  0.991685   
163   {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.973474   

      neg_dist  
1215  0.003231  
762   0.004358  
33    0.008315  
1255  0.008315  
163   0.026526

In [48]:
most_by_correct(preds_df, 'neg')

candidate_task  \
1045  Another aspect of this problem is prioritizati...   
121   35. "Stewardesses" is the longest word that is...   
83    5. A shark is the only fish that can blink wit...   
88                        building is an American flag.   
990   Common sense is the collection of prejudices a...   

                      prob_dist label  \
1045  <ProbDist with 2 samples>   neg   
121   <ProbDist with 2 samples>   neg   
83    <ProbDist with 2 samples>   neg   
88    <ProbDist with 2 samples>   neg   
990   <ProbDist with 2 samples>   neg   

                                               features pred  pos_dist  \
1045  {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  0.013304   
121   {'VERB': 3, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.015950   
83    {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.016720   
88    {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.019714   
990   {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  neg  0.022955   

      neg_dist  
1045  0.986696  
121   0.984050  
83    0.983280  
88    0.980286  
990   0.977045

In [49]:
most_by_correct(preds_df, 'pos', is_correct=False)

candidate_task                  prob_dist  \
380  maine impossible to get to...next option?  <ProbDist with 2 samples>   
191         Start project  with Elene for Mark  <ProbDist with 2 samples>   

    label                                           features pred  pos_dist  \
380   pos  {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  neg  0.330498   
191   pos  {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  neg  0.480333   

     neg_dist  
380  0.669502  
191  0.519667

In [50]:
most_by_correct(preds_df, 'neg', is_correct=False)

candidate_task  \
1282                                            use the   
529                                               Visit   
1091  end, after court hearings on the issue, the Dr...   
396   ENA Origination Support (Getting info from Tyc...   
1121  late Tuesday accused 29 Enron officers and dir...   

                      prob_dist label  \
1282  <ProbDist with 2 samples>   neg   
529   <ProbDist with 2 samples>   neg   
1091  <ProbDist with 2 samples>   neg   
396   <ProbDist with 2 samples>   neg   
1121  <ProbDist with 2 samples>   neg   

                                               features pred  pos_dist  \
1282  {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.963347   
529   {'VERB': 1, 'PARENT_DEP_ROOT': 1, 'PARENT_POS_...  pos  0.960683   
1091  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.955679   
396   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...  pos  0.949537   
1121  {'VERB': 1, 'FOLLOWING_POS_NUM': 1, 'FOLLOWING...  pos  0.942597   

      neg_dist  
1282  0.036653  
529   0.039317  
1091  0.044321  
396   0.050463  
1121  0.057403

Final `get_sample_predictions` goal: get most uncertain

`most_uncertain = np.argsort(np.abs(probs -0.5))[:4]`

Basically, we're going to sort by probabilities closest to 0.5 (on either side)

#### most_uncertain(preds_df, label, n=5)

In [51]:
def most_uncertain(preds_df, label, n=5):
    dist_column = label + '_dist'
    return preds_df.iloc[np.argsort(np.abs(preds_df[dist_column]-0.5))][:n]

In [52]:
most_uncertain(preds_df, 'pos', 10)

candidate_task  \
965   Currently, most of the trades are between secu...   
573   Davis has taken heat for locking up too much e...   
614   I am learning peacefulness, lying by myself qu...   
721   about Australian markets and was greatly impre...   
1411  Incentive fee (in main contract) defined to on...   
1317  trader I see an opportunity to sell soybeans a...   
901   The oldest of the insurance-related, exchange-...   
116   31. The microwave was invented after a researc...   
904   Initially, there was substantial interest in t...   
909   so far this year, he said. Last year, there wa...   

                      prob_dist label  \
965   <ProbDist with 2 samples>   neg   
573   <ProbDist with 2 samples>   neg   
614   <ProbDist with 2 samples>   neg   
721   <ProbDist with 2 samples>   neg   
1411  <ProbDist with 2 samples>   neg   
1317  <ProbDist with 2 samples>   neg   
901   <ProbDist with 2 samples>   neg   
116   <ProbDist with 2 samples>   neg   
904   <ProbDist with 2 samples>   neg   
909   <ProbDist with 2 samples>   neg   

                                               features pred  pos_dist  \
965   {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.500492   
573   {'VERB': 6, 'FOLLOWING_POS_VERB': 3, 'FOLLOWIN...  neg  0.498548   
614   {'VERB': 3, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  pos  0.501745   
721   {'VERB': 1, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...  neg  0.496814   
1411  {'VERB': 2, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.504848   
1317  {'VERB': 5, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.505635   
901   {'VERB': 2, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  neg  0.492510   
116   {'VERB': 3, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  neg  0.491751   
904   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  neg  0.491579   
909   {'VERB': 3, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.512387   

      neg_dist  
965   0.499508  
573   0.501452  
614   0.498255  
721   0.503186  
1411  0.495152  
1317  0.494365  
901   0.507490  
116   0.508249  
904   0.508421  
909   0.487613

In [53]:
most_uncertain(preds_df, 'neg', 10)

candidate_task  \
965   Currently, most of the trades are between secu...   
573   Davis has taken heat for locking up too much e...   
614   I am learning peacefulness, lying by myself qu...   
721   about Australian markets and was greatly impre...   
1411  Incentive fee (in main contract) defined to on...   
1317  trader I see an opportunity to sell soybeans a...   
901   The oldest of the insurance-related, exchange-...   
116   31. The microwave was invented after a researc...   
904   Initially, there was substantial interest in t...   
909   so far this year, he said. Last year, there wa...   

                      prob_dist label  \
965   <ProbDist with 2 samples>   neg   
573   <ProbDist with 2 samples>   neg   
614   <ProbDist with 2 samples>   neg   
721   <ProbDist with 2 samples>   neg   
1411  <ProbDist with 2 samples>   neg   
1317  <ProbDist with 2 samples>   neg   
901   <ProbDist with 2 samples>   neg   
116   <ProbDist with 2 samples>   neg   
904   <ProbDist with 2 samples>   neg   
909   <ProbDist with 2 samples>   neg   

                                               features pred  pos_dist  \
965   {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.500492   
573   {'VERB': 6, 'FOLLOWING_POS_VERB': 3, 'FOLLOWIN...  neg  0.498548   
614   {'VERB': 3, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  pos  0.501745   
721   {'VERB': 1, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...  neg  0.496814   
1411  {'VERB': 2, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...  pos  0.504848   
1317  {'VERB': 5, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...  pos  0.505635   
901   {'VERB': 2, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  neg  0.492510   
116   {'VERB': 3, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...  neg  0.491751   
904   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...  neg  0.491579   
909   {'VERB': 3, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...  pos  0.512387   

      neg_dist  
965   0.499508  
573   0.501452  
614   0.498255  
721   0.503186  
1411  0.495152  
1317  0.494365  
901   0.507490  
116   0.508249  
904   0.508421  
909   0.487613

Turns out, both `label`s return the same list

## Writing: aggregate_features

how to correlate features to (in)correctness

Example set of feature dictionaries...

In [54]:
most_incorrect[['features']]

features
1282  {'VERB': 1, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...
529   {'VERB': 1, 'PARENT_DEP_ROOT': 1, 'PARENT_POS_...
1091  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...
396   {'VERB': 1, 'FOLLOWING_POS_NOUN': 1, 'FOLLOWIN...
1121  {'VERB': 1, 'FOLLOWING_POS_NUM': 1, 'FOLLOWING...
1031  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...
926   {'VERB': 3, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...
167   {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...
752   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...
196   {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...
872   {'VERB': 2, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...
638   {'VERB': 2, 'FOLLOWING_POS_ADP': 2, 'FOLLOWING...
923   {'VERB': 1, 'FOLLOWING_POS_ADJ': 1, 'FOLLOWING...
1081  {'VERB': 2, 'FOLLOWING_POS_PART': 1, 'FOLLOWIN...
1305  {'VERB': 1, 'FOLLOWING_POS_PUNCT': 1, 'FOLLOWI...
1176  {'VERB': 4, 'FOLLOWING_POS_VERB': 1, 'FOLLOWIN...
1061  {'VERB': 1, 'FOLLOWING_POS_ADP': 1, 'FOLLOWING...
999   {'VERB': 3, 'FOLLOWING_POS_ADV': 1, 'FOLLOWING...
1118  {'VERB': 2, 'FOLLOWING_POS_DET': 1, 'FOLLOWING...
1058  {'VERB': 1, 'FOLLOWING_POS_PROPN': 1, 'FOLLOWI...

Convert one feature dictionary to dataframe...

In [55]:
most_incorrect[['features']].iloc[1].values[0]

{'PARENT_DEP_ROOT': 1, 'PARENT_POSTAG_VB': 1, 'PARENT_POS_VERB': 1, 'VERB': 1}

In [56]:
f_dict = pd.DataFrame.from_dict(most_incorrect[['features']].iloc[1].values[0], orient='index').reset_index()
f_dict

index  0
0              VERB  1
1   PARENT_DEP_ROOT  1
2   PARENT_POS_VERB  1
3  PARENT_POSTAG_VB  1

In [57]:
f_dict.columns = ['feature', 'count']
f_dict

feature  count
0              VERB      1
1   PARENT_DEP_ROOT      1
2   PARENT_POS_VERB      1
3  PARENT_POSTAG_VB      1

Aggregating two feature dataframes into one by feature occurrence count...

In [58]:
pd.concat([f_dict, f_dict]).groupby('feature').sum().reset_index()

feature  count
0   PARENT_DEP_ROOT      2
1  PARENT_POSTAG_VB      2
2   PARENT_POS_VERB      2
3              VERB      2

Putting it all together...

In [59]:
f_dicts = []
for i in range(len(most_incorrect)):
    f_dict = pd.DataFrame.from_dict(most_incorrect[['features']].iloc[i].values[0], orient='index').reset_index()
    f_dict.columns = ['feature', 'count']
    f_dicts.append(f_dict)
f_dicts

[               feature  count
 0                 VERB      1
 1    FOLLOWING_POS_DET      1
 2  FOLLOWING_POSTAG_DT      1
 3      PARENT_DEP_ROOT      1
 4      PARENT_POS_VERB      1
 5     PARENT_POSTAG_VB      1
 6       CHILD_DEP_DOBJ      1
 7        CHILD_POS_DET      1
 8      CHILD_POSTAG_DT      1,             feature  count
 0              VERB      1
 1   PARENT_DEP_ROOT      1
 2   PARENT_POS_VERB      1
 3  PARENT_POSTAG_VB      1,                 feature  count
 0                  VERB      1
 1   FOLLOWING_POS_PUNCT      1
 2    FOLLOWING_POSTAG_,      1
 3       PARENT_DEP_ROOT      1
 4       PARENT_POS_VERB      1
 5      PARENT_POSTAG_VB      1
 6       CHILD_DEP_PUNCT      1
 7       CHILD_POS_PUNCT      1
 8        CHILD_POSTAG_,      1
 9        CHILD_DEP_PREP      1
 10        CHILD_POS_ADP      1
 11      CHILD_POSTAG_IN      1
 12       CHILD_DEP_CONJ      1
 13       CHILD_POS_NOUN      1
 14      CHILD_POSTAG_NN      1,                feature  count
 0     

In [60]:
pd.concat(f_dicts).groupby('feature').sum().sort_values('count', ascending=False).reset_index()

feature  count
0                    VERB     31
1         PARENT_POS_VERB     28
2         PARENT_DEP_ROOT     27
3          CHILD_POS_NOUN     19
4        PARENT_POSTAG_VB     18
5          CHILD_DEP_DOBJ     14
6         CHILD_POS_PUNCT     13
7         CHILD_DEP_PUNCT     13
8          CHILD_POS_VERB     12
9           CHILD_POS_ADP     12
10        CHILD_POSTAG_IN     12
11        CHILD_POSTAG_NN     11
12          CHILD_DEP_AUX      9
13         CHILD_DEP_PREP      9
14        CHILD_POS_PROPN      8
15       CHILD_POSTAG_NNS      8
16       CHILD_POSTAG_NNP      8
17         CHILD_POS_PART      7
18        CHILD_POSTAG_VB      6
19      PARENT_POSTAG_VBD      6
20      FOLLOWING_POS_DET      6
21         CHILD_POSTAG_,      6
22     CHILD_DEP_NPADVMOD      6
23        CHILD_DEP_NSUBJ      6
24    FOLLOWING_POSTAG_DT      6
25         CHILD_POSTAG_.      5
26    FOLLOWING_POSTAG_IN      5
27      FOLLOWING_POS_ADP      5
28        CHILD_POSTAG_TO      5
29        CHILD_POSTAG_RB      4
..                    ...    ...
64        CHILD_POSTAG_DT      1
65      CHILD_DEP_AUXPASS      1
66        CHILD_POSTAG_CD      1
67       CHILD_POSTAG_VBG      1
68       PARENT_POSTAG_IN      1
69       PARENT_DEP_XCOMP      1
70    CHILD_DEP_NSUBJPASS      1
71        PARENT_DEP_PREP      1
72         PARENT_POS_ADP      1
73       CHILD_POSTAG_ADD      1
74        CHILD_POSTAG_UH      1
75       CHILD_POSTAG_VBD      1
76      FOLLOWING_POS_NUM      1
77    FOLLOWING_POSTAG_CD      1
78  FOLLOWING_POSTAG_PRP$      1
79     FOLLOWING_POSTAG_.      1
80     FOLLOWING_POSTAG_,      1
81            CHILD_POS_X      1
82    FOLLOWING_POSTAG_NN      1
83     CHILD_DEP_COMPOUND      1
84   FOLLOWING_POSTAG_NNS      1
85         CHILD_DEP_INTJ      1
86          CHILD_DEP_NEG      1
87   FOLLOWING_POSTAG_VBG      1
88   FOLLOWING_POSTAG_VBN      1
89          CHILD_POS_NUM      1
90         CHILD_POS_INTJ      1
91          CHILD_POS_DET      1
92    FOLLOWING_POS_CCONJ      1
93    FOLLOWING_POSTAG_CC      1

[94 rows x 2 columns]

#### aggregate_features(preds_df, n=0)

In [61]:
def aggregate_features(preds_df, n=0):
    f_dicts = []
    for i in range(len(most_incorrect)):
        f_dict = pd.DataFrame.from_dict(preds_df[['features']].iloc[i].values[0], orient='index').reset_index()
        f_dict.columns = ['feature', 'count']
        f_dicts.append(f_dict)
    
    agg = pd.concat(f_dicts).groupby('feature').sum().sort_values('count', ascending=False).reset_index()
    
    if n <= 0:
        return agg
    else:
        return agg[:n]

In [62]:
aggregate_features(most_incorrect)

feature  count
0                    VERB     31
1         PARENT_POS_VERB     28
2         PARENT_DEP_ROOT     27
3          CHILD_POS_NOUN     19
4        PARENT_POSTAG_VB     18
5          CHILD_DEP_DOBJ     14
6         CHILD_POS_PUNCT     13
7         CHILD_DEP_PUNCT     13
8          CHILD_POS_VERB     12
9           CHILD_POS_ADP     12
10        CHILD_POSTAG_IN     12
11        CHILD_POSTAG_NN     11
12          CHILD_DEP_AUX      9
13         CHILD_DEP_PREP      9
14        CHILD_POS_PROPN      8
15       CHILD_POSTAG_NNS      8
16       CHILD_POSTAG_NNP      8
17         CHILD_POS_PART      7
18        CHILD_POSTAG_VB      6
19      PARENT_POSTAG_VBD      6
20      FOLLOWING_POS_DET      6
21         CHILD_POSTAG_,      6
22     CHILD_DEP_NPADVMOD      6
23        CHILD_DEP_NSUBJ      6
24    FOLLOWING_POSTAG_DT      6
25         CHILD_POSTAG_.      5
26    FOLLOWING_POSTAG_IN      5
27      FOLLOWING_POS_ADP      5
28        CHILD_POSTAG_TO      5
29        CHILD_POSTAG_RB      4
..                    ...    ...
64        CHILD_POSTAG_DT      1
65      CHILD_DEP_AUXPASS      1
66        CHILD_POSTAG_CD      1
67       CHILD_POSTAG_VBG      1
68       PARENT_POSTAG_IN      1
69       PARENT_DEP_XCOMP      1
70    CHILD_DEP_NSUBJPASS      1
71        PARENT_DEP_PREP      1
72         PARENT_POS_ADP      1
73       CHILD_POSTAG_ADD      1
74        CHILD_POSTAG_UH      1
75       CHILD_POSTAG_VBD      1
76      FOLLOWING_POS_NUM      1
77    FOLLOWING_POSTAG_CD      1
78  FOLLOWING_POSTAG_PRP$      1
79     FOLLOWING_POSTAG_.      1
80     FOLLOWING_POSTAG_,      1
81            CHILD_POS_X      1
82    FOLLOWING_POSTAG_NN      1
83     CHILD_DEP_COMPOUND      1
84   FOLLOWING_POSTAG_NNS      1
85         CHILD_DEP_INTJ      1
86          CHILD_DEP_NEG      1
87   FOLLOWING_POSTAG_VBG      1
88   FOLLOWING_POSTAG_VBN      1
89          CHILD_POS_NUM      1
90         CHILD_POS_INTJ      1
91          CHILD_POS_DET      1
92    FOLLOWING_POS_CCONJ      1
93    FOLLOWING_POSTAG_CC      1

[94 rows x 2 columns]

In [63]:
aggregate_features(most_incorrect, 10)

feature  count
0              VERB     31
1   PARENT_POS_VERB     28
2   PARENT_DEP_ROOT     27
3    CHILD_POS_NOUN     19
4  PARENT_POSTAG_VB     18
5    CHILD_DEP_DOBJ     14
6   CHILD_POS_PUNCT     13
7   CHILD_DEP_PUNCT     13
8    CHILD_POS_VERB     12
9     CHILD_POS_ADP     12

In [64]:
aggregate_features(most_correct, 10)

feature  count
0             VERB     35
1  PARENT_POS_VERB     27
2   CHILD_POS_NOUN     26
3  PARENT_DEP_ROOT     25
4  CHILD_POSTAG_NN     23
5  CHILD_DEP_PUNCT     18
6  CHILD_POS_PUNCT     18
7  CHILD_DEP_NSUBJ     15
8   CHILD_DEP_DOBJ     14
9   CHILD_POSTAG_.     13